### PR TITLE
[3D] texture cubemap GL & skybox 

### PR DESCRIPTION
**Pull request by PatriceJiang**
OpenGL 
- add textureCube implementation
- enable skybox & skybox testcase

### CODE REVIEW

File: cocos/renderer/backend/Types.h [hunk 1]
```diff
@@ -245,4 +245,14 @@ struct AttributeBindInfo
     int         type        = 0;
 };
 
+enum class TextureCubeFace : int32_t
```
Comment (minggo) [inline cocos/renderer/backend/Types.h]: int32_t -> uint32

File: cocos/renderer/backend/Device.h [hunk 2]
```diff
@@ -36,8 +36,10 @@ class Device : public cocos2d::Ref
     virtual CommandBuffer* newCommandBuffer() = 0;
     // Create a buffer, not auto released.
     virtual Buffer* newBuffer(uint32_t size, BufferType type, BufferUsage usage) = 0;
-    // Create a texture, not auto released.
-    virtual Texture* newTexture(const TextureDescriptor& descriptor) = 0;
+    // Create a 2d texture, not auto released.
+    virtual Texture* newTexture2D(const TextureDescriptor& descriptor) = 0;
+    // Create a cube texture, not auto released
+    virtual Texture* newTextureCube(const TextureDescriptor& descriptor) = 0;
```
Comment (minggo) [inline cocos/renderer/backend/Device.h]: i think only one interface is enough

File: cocos/renderer/backend/Texture.h [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
 #include "Types.h"
 #include "base/CCRef.h"
+#include "renderer/CCTexture2D.h"
```
Comment (minggo) [inline cocos/renderer/backend/Texture.h]: backend do not need to include engine codes

File: cocos/renderer/backend/Texture.h [hunk 4]
```diff
@@ -10,27 +11,31 @@ struct TextureDescriptor
 {
     TextureType textureType = TextureType::TEXTURE_2D;
     TextureFormat textureFormat = TextureFormat::R8G8B8A8;
-    TextureUsage textureUsage = TextureUsage::READ;
-    uint32_t width = 0;
-    uint32_t height = 0;
+    TextureUsage textureUsage = TextureUsage::READ;;
```
Comment (minggo) [inline cocos/renderer/backend/Texture.h]: two `;;`

File: cocos/renderer/CCTextureCube.cpp [hunk 5]
```diff
@@ -67,7 +70,7 @@ unsigned char* getImageData(Image* img, Texture2D::PixelFormat&  ePixFmt)
         {
             // Convert "RRRRRRRRRGGGGGGGGBBBBBBBBAAAAAAAA" to "RRRRRGGGGGGBBBBB"
             inPixel32 = (unsigned int*)img->getData();
-            pTmpData = new (std::nothrow) unsigned char[nWidth * nHeight * 2];
+            pTmpData = (unsigned char *)malloc(nWidth * nHeight * 2);
```
Comment (minggo) [inline cocos/renderer/CCTextureCube.cpp:73]: Why change to malloc?

File: cocos/renderer/backend/Texture.h [hunk 6]
```diff
@@ -11,39 +11,59 @@ struct TextureDescriptor
...
     TextureUsage textureUsage = TextureUsage::READ;
-    uint32_t width = 0;
-    uint32_t height = 0;
     uint32_t depth = 0;
     bool compressed = false;
+    SamplerDescriptor samplerDescriptor;    
     
-    SamplerDescriptor samplerDescriptor;
+    uint32_t width = 0;
+    uint32_t height = 0;
```
Comment (minggo) [inline cocos/renderer/backend/Texture.h:15]: Why need to change the order?

File: cocos/renderer/backend/Texture.h [hunk 7]
```diff
@@ -11,39 +11,59 @@ struct TextureDescriptor
...
+
+};
+
+class TextureCubemap : public Texture
+{
+public:
+    virtual void updateFaceData(TextureCubeFace side, int size, void *data) = 0;
+protected:
+    TextureCubemap(const TextureDescriptor& descriptor);
+    uint32_t _size = 0;
```
Comment (minggo) [inline cocos/renderer/backend/Texture.h:66]: size per face?

File: cocos/renderer/backend/opengl/TextureGL.cpp [hunk 8]
```diff
@@ -235,169 +413,74 @@ void TextureGL::updateSubData(unsigned int xoffset, unsigned int yoffset, unsign
...
+    toGLSamplerDescriptor(sampler, true);
+    glActiveTexture(GL_TEXTURE0);
+    glBindTexture(GL_TEXTURE_CUBE_MAP, _texture);
+    
+    glTexParameteri(GL_TEXTURE_CUBE_MAP, GL_TEXTURE_MIN_FILTER, _minFilterGL);
+    glTexParameteri(GL_TEXTURE_CUBE_MAP, GL_TEXTURE_MAG_FILTER, _magFilterGL);
+    glTexParameteri(GL_TEXTURE_CUBE_MAP, GL_TEXTURE_WRAP_S, _sAddressModeGL);
+    glTexParameteri(GL_TEXTURE_CUBE_MAP, GL_TEXTURE_WRAP_T, _tAddressModeGL);
+    
+    glActiveTexture(GL_TEXTURE0);
```
Comment (minggo) [inline cocos/renderer/backend/opengl/TextureGL.cpp]: It is active above, so don't have to active again.

File: cocos/renderer/backend/opengl/TextureGL.cpp [hunk 9]
```diff
@@ -235,169 +413,74 @@ void TextureGL::updateSubData(unsigned int xoffset, unsigned int yoffset, unsign
...
+    int i = static_cast<int>(side);
+    glTexImage2D(GL_TEXTURE_CUBE_MAP_POSITIVE_X + i,
+        0,                  // level
+        GL_RGBA,            // internal format
+        size,              // width
+        size,             // height
+        0,                  // border
+        _internalFormat,            // format
+        _type,   // type
+        data);              // pixel data
```
Comment (minggo) [inline cocos/renderer/backend/opengl/TextureGL.cpp:482]: wrong indent of the comment

File: cocos/renderer/backend/opengl/TextureGL.h [hunk 10]
```diff
@@ -5,38 +5,54 @@
 
 CC_BACKEND_BEGIN
 
-class TextureGL : public Texture
+class TextureInfoGL
```
Comment (minggo) [inline cocos/renderer/backend/opengl/TextureGL.h]: why need this class?

File: cocos/renderer/backend/opengl/TextureGL.h [hunk 11]
```diff
@@ -5,38 +5,52 @@
...
+};
+
+class TextureCubeGL: public backend::TextureCubemap
+{
+public:
+    TextureCubeGL(const TextureDescriptor& descriptor);
+    ~TextureCubeGL();
+    void apply(int index) const;
+    virtual void updateSamplerDescriptor(const SamplerDescriptor &sampler) override;
+    virtual void updateFaceData(TextureCubeFace side, void *data) override;
```
Comment (minggo) [inline cocos/renderer/backend/opengl/TextureGL.h:53]: Virtual functions should be declared before other functions: `apply`.

**Review by Mee-gu — APPROVED** [review]

**Review by minggo — APPROVED** [review]

**Pull request merged**